Switch to labels

okdas opened the pull request:
# Redesign metrics to use labels instead of metric name embedding

This PR completely redesigns the metric structure to follow Prometheus best practices by using labels instead of embedding information in metric names.

## Key Changes

### Metric Structure
- Consolidated metrics into a single `vultr_account_bandwidth_value` metric with labels
- Replaced product-specific billing metrics with generic `billing_cost_usd` and `billing_units` metrics
- Renamed `kubernetes_node` to `kubernetes_node_pool_nodes` for clarity
- Added `block_type` label to Block Storage metrics

### API Improvements
- Added pagination support to all API calls (Block Storage, Kubernetes, Load Balancer, Reserved IPs)
- Increased billing API page size to 500 items to get reporting on large lists
- Improved error logging with detailed error messages
- Removed excessive debug logging statements

### Documentation
- Added more metrics documentation in README
- Added useful PromQL query examples for common use cases
- Added examples for cost analysis, resource usage, and filtering

## Testing
- Updated test expectations to match new metric format
- All tests passing
- Manually verified complete resource collection with pagination

DazWilkin commented:
Switching from hard-coding values into the metric name (labels) is a good idea, thank you!

DazWilkin commented on collector/invoice_item.go:
```diff
@@ -1,22 +1,38 @@
+	currentUnits     map[string]float64 // Maps unit_type to total units
+	currentUnitPrice map[string]float64 // Maps unit_type to latest unit price
+	currentTotal     float64            // Running total for all items
+	description      string             // Description of the item
```
Using `description` and `product` is duplicative. 

I wonder whether this value may be changed over time? Please consider dropping `description`.

Should we trivially normalize e.g. `Load Balancer` to `load_balancer`?

Should we define a lookup for Products to ensure consistency if Vultr we e.g. to rename products?

okdas replied:
> Using description and product is duplicative.
> I wonder whether this value may be changed over time? Please consider dropping description.

They are both needed. `product` is more like a category (LB, Snapshot, etc..), while `description` allows us to get metrics per particular resource (server1, server2, lb1, lb2):
```
vultr_billing_units{description="50 GB Block Storage New Jersey [pvc-959d99cbc262496c]",product="Block Storage",unit_price="0.007440",unit_type="hours"} 118
vultr_billing_units{description="50 GB Block Storage New Jersey [pvc-ac3b32cce2c74e4a]",product="Block Storage",unit_price="0.007440",unit_type="hours"} 118
vultr_billing_units{description="50 GB NVMe Block Storage (New Jersey)  (pvc-8af90c5364324118)",product="Block Storage",unit_price="0.007440",unit_type="hours"} 648
vultr_billing_units{description="y.y.y.y (16384 MB) [protocol-k8s-nj-6c-16g-08a0f846ad80]",product="Vultr Cloud Compute",unit_price="0.119000",unit_type="hours"} 1
vultr_billing_units{description="x.x.x.x (16384 MB) [protocol-k8s-nj-6c-16g-0f5b5656de99]",product="Vultr Cloud Compute",unit_price="0.119000",unit_type="hours"} 30
```

> Should we trivially normalize e.g. Load Balancer to load_balancer?

I'd keep these values as from the API as they are. I trust Vultr that if they break the API, they will at least communicate that to users.

> Should we define a lookup for Products to ensure consistency if Vultr we e.g. to rename products?

Sorry - can you please elaborate? I'm thinking if they change the name in one place, it would also be changed in another? The current logic is not opinionated on what resources are. E.g. if Vultr creates a new product, it will be discovered by the exporter.

DazWilkin replied:
Thanks for providing the explanation of `description`; I've not used Vultr sufficiently to be exposed to this level of detail.

I (still) *really* dislike this composite label; I see the value in its attributes but retaining it as a string increases the number of time-series quite significantly:

+ each distinct IP address with Vultr Cloud Compute
+ each PVC name with Block Storage Unit).

I think we should provide the user with a mechanism (flag?) to:

1. Disable `description` label
2. Leave `description` label as-is
3. Optional: parse `description` label into sub-labels (e.g. `size`, `type`, `location` for Block Storage

I'll defer this to you but that's my preference.

okdas replied:
@DazWilkin 

> I (still) really dislike this composite label

Me too, but that is a flaw of Vultr API - I wish they had a better structure. Parsing is likely going to worsen it - more potential issues. I'd merge it as is, and if the need becomes apparent - add a parsing feature, too. Sounds good?

> I see the value in its attributes but retaining it as a string increases the number of time-series quite significantl

We can provide a way to optionally turn on/off description label and aggregate something like `sum() by (product)` when it's turned off. Do you have preference of default value?

Imo should be on by default:
- `description` is not a [high-cardinality label](https://chronosphere.io/learn/what-is-high-cardinality/) in my view.
- metrics with `description` label have no churning - they only appear/disappear when actual resources provisioned/removed. Happens rarely, even on volatile k8s clusters.
- If the infrastructure is small, there will be only a handful of time-series.
- When there are many resources deployed, and there are thousands of time-series, I would assume the observability stack is large enough for it not to be an issue. Just checked a small `ingress-nginx` deployment and it had almost 6k time series. It's not that much at all.

okdas commented:
@DazWilkin thanks for the review - I'll get back to this within the next few days.

okdas commented on collector/bandwidth.go:
```diff
@@ -14,45 +13,30 @@ var (
+		Value: prometheus.NewDesc(
+			prometheus.BuildFQName(s.Namespace, "account_bandwidth", "value"),
+			"Bandwidth metric value",
+			[]string{"period", "metric", "unit"},
```
> However using metric as a label name is confusing, please consider type or other name that's more reflective of the Bandwidth fields.

Agreed, will change.

> Time-related values are hard. Do we lose too much information by replacing the periods' timestamps (start|end) with a string enum?

I don’t see what information we lose. Prometheus (or any other monitoring stack/TSDB, I prefer VictoriaMetrics by far) regularly pulls data from the exporter, providing a datapoint for each pull. This allows us to query historical data if needed. I believe `start`/`end` are useful outside of TSDB storage contexts, such as when we need to create a detailed CSV with all expenses.

One technique I’ve observed in some projects is casting datatime into a timestamp and exporting this timestamp as a separate time-series. So we can calculate the delta between `start` and `end` to export this value. This approach creates a single timeseries that enables us to calculate both the start and end values. Given this metric comes from [pending_charges](https://www.vultr.com/api/#tag/billing/operation/pending-charges) endpoint, what are the potential uses for it? Seems like these values are not even documented, hmm.

DazWilkin replied:
My concern is that today's `current` is yesterday's `projected` and tomorrow's `previous`.

Is the user ever going to want to query across `period="current"` and be confused by the results?

okdas replied:
@DazWilkin to be frank, I, personally, am only interested in `period="current"` as it provides everything: historical data, current and allows extrapolating. It is intuitive to me to use the exporter that way.

I kept all three because that's what Vultr API exposes and it doesn't cost much to save that data to TSDB.

Seems like `previous` changes only once a month:
<img width="915" alt="image" src="https://github.com/user-attachments/assets/ccc37236-19fd-4949-9584-40f8d6382c8b" />

We can add documentation to clear up the confusion.

okdas commented:
@DazWilkin ready for another round 🫡 

In addition to requested change:
- Renamed metrics:
    - billing_total → billing_cost_usd to clearly indicate USD currency
    - kubernetes_node → kubernetes_node_pool_nodes for better clarity
- Added pagination support to all API calls with PerPage=100 or PerPage=500 to ensure all resources are properly collected
- Added block_type label to Block Storage metrics to enable filtering by storage type
- Improved error logging by replacing log.Info with log.Error and including error details
- Removed excessive debug logging statements that were cluttering logs

DazWilkin commented:
Thanks for making the requested changes.

Thanks also for the additional changes but please keep additional improvements to new PRs to make reviewing easier.

I'm still concerned about `current`,`previous`, `projected` and the `description`  label but reluctant to prolong this PR.

okdas commented:
@DazWilkin Thank you!

> keep additional improvements to new PRs to make reviewing easier

Will do. I apologize for the number of changes introduced.

I hope I provided decent solutions for concerns, in short, they are:
- better documentation for current/previous/projected.
- flag to turn off `description` tag and aggregate when it is turned off. I'd keep it on by default, but let me know if you feel otherwise.

DazWilkin commented:
No need to apologize but thank you. 

Thank you for all your contributions.

I don't want to labor this PR too much longer, please make the changes based on it discussion and I'll approve. 

I'm going into the weekend so my response may be delayed.
